File: backend/app/services/profile_scheduler_service.py
"""Profile scheduler — automatic profile switching based on time-of-day rules.

Checks the ``profile_schedules`` table every 60 seconds.  When the current
day/time matches a schedule, the associated profile is activated — but only
if the system is NOT in panic mode, NOT in an alert-triggered profile switch,
and NOT in quiet hours.

Priority order (highest to lowest):
  1. Panic mode (always wins)
  2. Alert-triggered profile (safety)
  3. Quiet hours (explicit user silence preference)
  4. Profile schedule (automation convenience)
  5. Manual profile selection (default)
"""

from __future__ import annotations

import asyncio
import logging
from datetime import datetime, time, timezone

import aiosqlite

logger = logging.getLogger(__name__)


class ProfileSchedulerService:
    def __init__(self, db: aiosqlite.Connection) -> None:
        self._db = db
        self._task: asyncio.Task | None = None
        self._running = False
        # Track which schedule is currently applied so we don't re-activate
        # every 60 seconds.
        self._active_schedule_id: str | None = None
        self._activate_profile_fn = None
        # Callbacks to check higher-priority overrides
        self._is_panic_fn = None
        self._is_alert_active_fn = None
        self._is_quiet_hours_fn = None

    def set_activate_fn(self, fn) -> None:
        """Set the callback used to activate a profile by ID."""
        self._activate_profile_fn = fn

    def set_panic_check_fn(self, fn) -> None:
        """Set callback that returns True if system is in panic mode."""
        self._is_panic_fn = fn

    def set_alert_active_check_fn(self, fn) -> None:
        """Set callback that returns True if an alert-triggered profile is active."""
        self._is_alert_active_fn = fn

    def set_quiet_hours_check_fn(self, fn) -> None:
        """Set callback that returns True if quiet hours are currently active."""
        self._is_quiet_hours_fn = fn

    async def start(self) -> None:
        self._running = True
        self._task = asyncio.create_task(self._loop())

    async def stop(self) -> None:
        self._running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass

    async def _loop(self) -> None:
        while self._running:
            try:
                await self._check_schedule()
            except asyncio.CancelledError:
                raise
            except Exception:
                logger.exception("Profile schedule check failed")
            await asyncio.sleep(60)

    async def _check_schedule(self) -> None:
        if self._activate_profile_fn is None:
            return

        # Check higher-priority overrides
        if self._is_panic_fn and self._is_panic_fn():
            return
        if self._is_alert_active_fn and self._is_alert_active_fn():
            return
        if self._is_quiet_hours_fn and self._is_quiet_hours_fn():
            return

        schedules = await self._load_schedules()
        matched = _find_active_schedule(schedules)

        if matched is None:
            # No schedule matches — clear tracking
            if self._active_schedule_id is not None:
                self._active_schedule_id = None
            return

        schedule_id = matched["id"]
        profile_id = matched["profile_id"]

        # Already on this schedule — nothing to do
        if self._active_schedule_id == schedule_id:
            return

        self._active_schedule_id = schedule_id

        try:
            await self._activate_profile_fn(profile_id)
            logger.info(
                "Profile schedule: activated profile %s (schedule %s)",
                profile_id, schedule_id,
            )
        except Exception:
            logger.exception(
                "Profile schedule: failed to activate profile %s", profile_id
            )

    async def _load_schedules(self) -> list[dict]:
        cursor = await self._db.execute(
            "SELECT id, profile_id, start_time, end_time, days_of_week, "
            "timezone, enabled, created_at "
            "FROM profile_schedules WHERE enabled = 1"
        )
        rows = await cursor.fetchall()
        return [
            {
                "id": r[0],
                "profile_id": r[1],
                "start_time": r[2],
                "end_time": r[3],
                "days_of_week": r[4],
                "timezone": r[5],
                "enabled": bool(r[6]),
                "created_at": r[7],
            }
            for r in rows
        ]


def _find_active_schedule(
    schedules: list[dict],
    now: datetime | None = None,
) -> dict | None:
    """Find the best matching schedule for the current local time.

    For overlapping schedules: most specific (fewest days) wins.
    If tied, most recently created wins.
    """
    if now is None:
        now = datetime.now(timezone.utc)

    current_day = now.weekday()  # 0=Monday
    current_time = now.strftime("%H:%M")

    matching = []
    for schedule in schedules:
        if not schedule.get("enabled", True):
            continue
        days = [int(d) for d in schedule["days_of_week"].split(",") if d.strip()]
        if current_day not in days:
            continue

        start = schedule["start_time"]
        end = schedule["end_time"]

        # Handle overnight spans (e.g., 22:00 to 06:00)
        if start <= end:
            if start <= current_time < end:
                matching.append(schedule)
        else:  # overnight
            if current_time >= start or current_time < end:
                matching.append(schedule)

    if not matching:
        return None

    # Most specific (fewest days) wins, then most recently created
    matching.sort(
        key=lambda s: (len(s["days_of_week"].split(",")), s.get("created_at", ""))
    )
    return matching[0]

File: backend-cs/Services/ProfileSchedulerService.cs
using DriveChill.Models;

namespace DriveChill.Services;

/// <summary>
/// Background service that checks profile schedules every 60 seconds and
/// activates the matching profile — but only if the system is NOT in panic
/// mode, NOT in an alert-triggered profile switch, and NOT in quiet hours.
///
/// Priority order (highest to lowest):
///   1. Panic mode (always wins)
///   2. Alert-triggered profile (safety)
///   3. Quiet hours (explicit user silence preference)
///   4. Profile schedule (automation convenience)
///   5. Manual profile selection (default)
/// </summary>
public sealed class ProfileSchedulerService : BackgroundService
{
    private readonly DbService     _db;
    private readonly SettingsStore  _store;
    private readonly FanService    _fans;
    private readonly AlertService  _alerts;
    private readonly ILogger<ProfileSchedulerService> _log;

    /// <summary>Track which schedule is currently applied to avoid re-activating every 60s.</summary>
    private string? _activeScheduleId;

    public ProfileSchedulerService(
        DbService db,
        SettingsStore store,
        FanService fans,
        AlertService alerts,
        ILogger<ProfileSchedulerService> log)
    {
        _db     = db;
        _store  = store;
        _fans   = fans;
        _alerts = alerts;
        _log    = log;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        // Small delay so other services can initialise first
        await Task.Delay(2000, stoppingToken);

        while (!stoppingToken.IsCancellationRequested)
        {
            try
            {
                await CheckScheduleAsync(stoppingToken);
            }
            catch (OperationCanceledException) { throw; }
            catch (Exception ex)
            {
                _log.LogWarning(ex, "Profile schedule check failed");
            }

            await Task.Delay(60_000, stoppingToken);
        }
    }

    private async Task CheckScheduleAsync(CancellationToken ct)
    {
        // Higher-priority overrides: panic, alert-triggered, quiet hours
        if (_fans.IsInPanic)
            return;
        if (_alerts.HasActiveProfileSwitch)
            return;
        // Check quiet hours: if the SensorWorker's quiet hours logic has an active rule
        // we defer. We check via DbService if any quiet hours rule matches now.
        if (await IsQuietHoursActiveAsync(ct))
            return;

        var schedules = await _db.GetProfileSchedulesAsync(ct);
        var matched = FindActiveSchedule(schedules);

        if (matched == null)
        {
            if (_activeScheduleId != null)
                _activeScheduleId = null;
            return;
        }

        if (_activeScheduleId == matched.Id)
            return;

        // Activate the profile
        var profiles = _store.LoadProfiles().ToList();
        var profile = profiles.FirstOrDefault(p => p.Id == matched.ProfileId);
        if (profile == null)
        {
            _log.LogWarning("Profile schedule {ScheduleId}: profile {ProfileId} not found",
                matched.Id, matched.ProfileId);
            return;
        }

        _activeScheduleId = matched.Id;

        foreach (var p in profiles) p.IsActive = p.Id == matched.ProfileId;
        _store.SaveProfiles(profiles);
        _fans.SetCurves(profile.Curves);
        _log.LogInformation("Profile schedule: activated profile {ProfileId} (schedule {ScheduleId})",
            matched.ProfileId, matched.Id);
    }

    private async Task<bool> IsQuietHoursActiveAsync(CancellationToken ct)
    {
        var rules = await _db.GetQuietHoursAsync(ct);
        var now = DateTimeOffset.UtcNow;
        var dow = (int)now.DayOfWeek;
        // Convert .NET DayOfWeek (0=Sunday) to Python convention (0=Monday)
        dow = dow == 0 ? 6 : dow - 1;
        var currentTime = now.ToString("HH:mm");

        foreach (var rule in rules)
        {
            if (!rule.Enabled) continue;
            if (rule.DayOfWeek != dow) continue;

            if (string.Compare(rule.StartTime, rule.EndTime, StringComparison.Ordinal) <= 0)
            {
                if (string.Compare(currentTime, rule.StartTime, StringComparison.Ordinal) >= 0
                    && string.Compare(currentTime, rule.EndTime, StringComparison.Ordinal) < 0)
                    return true;
            }
            else
            {
                // Overnight span
                if (string.Compare(currentTime, rule.StartTime, StringComparison.Ordinal) >= 0
                    || string.Compare(currentTime, rule.EndTime, StringComparison.Ordinal) < 0)
                    return true;
            }
        }
        return false;
    }

    /// <summary>
    /// Find the best matching schedule for the current UTC time.
    /// Most specific (fewest days) wins; ties broken by most recently created.
    /// </summary>
    internal static ProfileScheduleRecord? FindActiveSchedule(List<ProfileScheduleRecord> schedules)
    {
        var utcNow = DateTimeOffset.UtcNow;

        var matching = new List<ProfileScheduleRecord>();

        foreach (var schedule in schedules)
        {
            if (!schedule.Enabled) continue;

            // Convert UTC now to the schedule's local timezone for comparison
            // Supports both IANA (e.g. "America/New_York") and Windows timezone IDs
            DateTimeOffset localNow;
            try
            {
                var tz = TimeZoneInfo.FindSystemTimeZoneById(schedule.Timezone ?? "UTC");
                localNow = TimeZoneInfo.ConvertTime(utcNow, tz);
            }
            catch (TimeZoneNotFoundException)
            {
                localNow = utcNow; // fall back to UTC if timezone is invalid
            }

            var dow = (int)localNow.DayOfWeek;
            // Convert .NET DayOfWeek (0=Sunday) to Python convention (0=Monday)
            dow = dow == 0 ? 6 : dow - 1;
            var currentTime = localNow.ToString("HH:mm");

            var days = schedule.DaysOfWeek.Split(',')
                .Select(d => d.Trim())
                .Where(d => int.TryParse(d, out _))
                .Select(int.Parse)
                .ToList();
            if (!days.Contains(dow)) continue;

            var start = schedule.StartTime;
            var end = schedule.EndTime;

            if (string.Compare(start, end, StringComparison.Ordinal) <= 0)
            {
                if (string.Compare(currentTime, start, StringComparison.Ordinal) >= 0
                    && string.Compare(currentTime, end, StringComparison.Ordinal) < 0)
                    matching.Add(schedule);
            }
            else
            {
                // Overnight span
                if (string.Compare(currentTime, start, StringComparison.Ordinal) >= 0
                    || string.Compare(currentTime, end, StringComparison.Ordinal) < 0)
                    matching.Add(schedule);
            }
        }

        if (matching.Count == 0) return null;

        // Most specific (fewest days) wins, then most recently created
        return matching
            .OrderBy(s => s.DaysOfWeek.Split(',').Length)
            .ThenByDescending(s => s.CreatedAt)
            .First();
    }
}
